### ISSUE TITLE
Exception localmarkets/packages - No such file or directory 

### ISSUE THREAD

**Issue by D1no** [NucleusIDE/meteor-live-update#9]
**TL;DR:**
The standard ``meteor create --example localmarket`` throws an exception after adding ``live-update``. Trying to read local packages even though none are added or created. See below.

```
➜  localmarket  meteor add nucleuside:live-update  
                                              
Changes to your project's package version selections:
                                              
kevohagan:ramda           added, version 0.13.0
nucleuside:live-update    added, version 0.3.0
nucleuside:transcompiler  added, version 0.0.1

                                              
nucleuside:live-update: Live update the changes in browser without full page
reload
➜  localmarket  meteor                           
[[[[[ /meteor/_tests/localmarket ]]]]]

=> Started proxy.                             
=> Started MongoDB.                           
W20150511-17:44:41.238(3)? (STDERR)           
W20150511-17:44:41.240(3)? (STDERR) /Users/macuser/.meteor/packages/meteor-tool/.1.1.3.1wysac9++os.osx.x86_64+web.browser+web.cordova/mt-os.osx.x86_64/dev_bundle/server-lib/node_modules/fibers/future.js:245
W20150511-17:44:41.241(3)? (STDERR) 						throw(ex);
W20150511-17:44:41.241(3)? (STDERR) 						      ^
W20150511-17:44:41.241(3)? (STDERR) Error: ENOENT, no such file or directory '/meteor/_tests/localmarket/packages'
W20150511-17:44:41.241(3)? (STDERR)     at Object.fs.readdirSync (fs.js:666:18)
W20150511-17:44:41.242(3)? (STDERR)     at new PackageCollector (packages/nucleuside:live-update/js/server/package_collector.js:25:1)
W20150511-17:44:41.242(3)? (STDERR)     at new CssCollector (packages/nucleuside:live-update/js/server/css_collector.js:19:1)
W20150511-17:44:41.242(3)? (STDERR)     at Package (packages/nucleuside:live-update/js/server/css_collector.js:94:1)
W20150511-17:44:41.242(3)? (STDERR)     at /meteor/_tests/localmarket/.meteor/local/build/programs/server/packages/nucleuside_live-update.js:465:4
W20150511-17:44:41.242(3)? (STDERR)     at /meteor/_tests/localmarket/.meteor/local/build/programs/server/packages/nucleuside_live-update.js:472:3
W20150511-17:44:41.242(3)? (STDERR)     at /meteor/_tests/localmarket/.meteor/local/build/programs/server/boot.js:222:10
W20150511-17:44:41.243(3)? (STDERR)     at Array.forEach (native)
W20150511-17:44:41.243(3)? (STDERR)     at Function._.each._.forEach (/Users/macuser/.meteor/packages/meteor-tool/.1.1.3.1wysac9++os.osx.x86_64+web.browser+web.cordova/mt-os.osx.x86_64/dev_bundle/server-lib/node_modules/underscore/underscore.js:79:11)
W20150511-17:44:41.243(3)? (STDERR)     at /meteor/_tests/localmarket/.meteor/local/build/programs/server/boot.js:117:5
=> Exited with code: 8
```

I love your approach to finally implement live reload of html resources, which is contrary to #7 clearly not on MDGs priority list. Developers learned to love Meteor, but (frontend) Designers - with a much tighter iteration cycle during development - are still left out (wait & compile - they hate it!).

I recently asked on the meteor forums for [Frontend Development Workflow Advice](https://forums.meteor.com/t/frontend-development-workflow-live-edit-slow-meteor/4111); no answers so far. I thought about developing a brackets or similar extension / package, than I stumbled over your package. Maybe you should reply and also consider posting your package to [crater.io](http://crater.io). Meteor certainly passed the point to get developers attention. Its time to address designers, then business cases.

Keep up the good work. If I have the opportunity // PS: Loved the [fresh tech video](http://www.youtube.com/watch?v=Q9M2YLiF-Q4) - not everyone belongs to the schizophrenic US way of public vs. private communication.

**Comment by channikhabra** [comment]
Hey,

Glad you liked this package.

This package is actually built for Nucleus(IDE) which is an in-browser IDE for Meteor (very much under development). This Live-update package works almost 100% for nucleus, but it has a dependency that it needs the source-code to update every time it updates (well, of course). This is not a problem in nucleus since nucleus is ran from right inside the app and has access to the complete source code of the app. But for external editors, it become a small hiccup.

Unlike tradition js apps, meteor has a tightly coupled watch process. The app gets notified that "something has changed" only after the build is finished (the compile phase as you said). So we can piggy-back on meteor's watch. That means, for usual editors, we need to have a separate process to watch for changes, and then somehow send those changes back to the app before meteor is done with its building. Otherwise this package don't add much value.

Now that being said, that video you saw is actually pretty old, new live-update package is much faster and robust than that video. Changes update almost instantaneously and it also supports less/sass. But it works only for nucleus. 

Because of this reason, I am trimming down this package to be a mere library. Read this [commit message](https://github.com/NucleusIDE/meteor-live-update/commit/cbdcc318875e16f549455d5c1321f60392dadb44). This makes it ideal for nucleus (since nucleus has direct instantaneous access to all changes), and allow writing external adaptors for other editors.

So if you wanna develop a brackets plugin, please go ahead with it. I have updated the readme with how you can do it and about what new changes I've made etc. Along with full-disclosure of all the taboos this package do to achieve its goal.

Nucleus is sort of complicated project with many moving parts (this package being one of them), so i can't really give enough time to every part of it and then make them 100% independent packages as well. Although I try. So as I said, I am trimming this package down to a mere library, and am developing another package which would watch for changes in filesystem and use this package to do the real work. 

I am not posting this package as an answer, because it is not a solution by itself. When the other package I am working on is done, then that might be an answer for your question. 

That all said, the error you see above is real and need to be taken care of. I'll close this issue when I actually resolve it.

**Comment by D1no** [comment]
Thanks for the concise answer. Of course, a meteor IDE is a great project, though I guess most designers would hate not to be able to stick to their IDE/Environment of choice just to get a certain feature (the problem brackets faces currently).

So I hope you give this library the required attention to have a solid solution for a lasting problem; than adding value on top of that. For designers "cursor position <-> to code" is certainly a great feature only feasible via an IDE integration.

If nucleus or brackets is the right choice for that, I don't know, but meteor needs live html injection.

I stare this issue and hope for a soon update.

**Comment by channikhabra** [comment]
Hey,

This error should not be happening anymore. 

I've also implemented the solution for a generic hot-loader for meteor. 

There are two things you need to install actually. [meteor-hot-loader](https://github.com/channikhabra/meteor-hot-loader) and [channikhabra:hot-loader](https://github.com/channikhabra/channikhabra-hot-loader)

I had to create a separate `meteor-hot-loader` command because of the issues we talked about last time. Actually yesterday I did implement what I suggested you, a meteor-only solution ([channikhabra:hot-loader-legacy](https://github.com/channikhabra/channikhabra-hot-loader-legacy)) for hot loading, but it's too slow since meteor start rebuilding on every change. For this i had to create a separate app that would do the injection. It's really easy to use, you don't need any configuration or nothing. Read the [this readme](https://github.com/channikhabra/meteor-hot-loader) for usage.

Please give it a shot some time.

**Comment by D1no** [comment]
Hey,

so I gave it a spin. Well, unfortunately the `meteor-hot-loader`throws an exception up on socket connection.

```
➜  localmarket  meteor-hot-loader
Meteor hot-loader rumbling on *:3333
meteor-hot-loader watching in /Users/macuser/development/apps/_tests/localmarket
A CLIENT CONNECTED
SyntaxError: Unexpected end of input
    at Object.parse (native)
    at parse (/usr/local/lib/node_modules/meteor-hot-loader/node_modules/body-parser/lib/types/json.js:84:17)
    at /usr/local/lib/node_modules/meteor-hot-loader/node_modules/body-parser/lib/read.js:102:18
    at done (/usr/local/lib/node_modules/meteor-hot-loader/node_modules/body-parser/node_modules/raw-body/index.js:250:14)
    at IncomingMessage.onEnd (/usr/local/lib/node_modules/meteor-hot-loader/node_modules/body-parser/node_modules/raw-body/index.js:296:7)
    at IncomingMessage.g (events.js:199:16)
    at IncomingMessage.emit (events.js:104:17)
    at _stream_readable.js:908:16
    at process._tickCallback (node.js:355:11)
Hot push failed socket hang up
```

Again, the plain `meteor create --example localmarket`example with
`sudo npm install -g meteor-hot-loader` (sudo seems a necessary evil to get the binary working)
`meteor add channikhabra:hot-loader`
than
`meteor-hot-loader` & `meteor`

**Comment by channikhabra** [comment]
Thanks for trying it out. Honestly I didn't test it much, I had to implement it twice in a weekend. I'll fix this bug and get back to you soon.

**Comment by channikhabra** [comment]
Hey Dino,

I've fixed that bug. Please give it a shot again. 

Moral of the story is, always use `Buffer.byteLength(dataString)` when setting Content-Length header. I was using `dataString.length`. Stupid me.

**Comment by D1no** [comment]
So, now it seems to be working. Don't forget to version increment your `package.json` of meteor-hot-loader, otherwise npm update can not see the change.

Did a new `sudo npm install -g meteor-hot-loader`, now it works as soon meteor-hot-loader is running.
```
➜  localmarket  meteor-hot-loader
Meteor hot-loader rumbling
meteor-hot-loader watching in /Users/macuser/development/apps/_tests/localmarket
A CLIENT CONNECTED
Hot-Pushed /Users/macuser/development/apps/_tests/localmarket/client/templates/home.html
Hot-Pushed /Users/macuser/development/apps/_tests/localmarket/client/templates/home.html
Hot-Pushed /Users/macuser/development/apps/_tests/localmarket/client/templates/home.html
```

I have currently 0 time on my hand. If I get the chance I would integrate the start-up of `meteor-hot-loader` into the meteor build process by defining a [Package.registerBuildPlugin](https://github.com/meteor/meteor/wiki/CompileStep-API-for-Build-Plugin-Source-Handlers) / [see Meteorchef (scroll down)](http://themeteorchef.com/recipes/writing-a-package/#Package.registerBuildPlugin) in `channikhabra:hot-loader` to dynamically check and/or start up the express server an a port relative to the one meteor is running (not fixed 3333) for multiple instances.

At some point a meteor only solution would be great; but for now the build process is killing it. Agree with you.

**Comment by channikhabra** [comment]
Time is a scarce resource for me as well. There are many things that need improvement in hot-loading process, that's why I haven't put it on meteor forums or elsewhere yet. 

You might have noticed a 1-2s delay in the saving the file and seeing the updates, that's because of the filesystem watcher. It can actually be improved to near instant. I hope the build plugins would be faster to notify of new changes in the file. I haven't really tried making a build plugin yet, but may be it can keep us from running two processes in two terminals (meteor and hot-loader). But I am afraid that build plugins might also suffer from same problem of meteor stopping everything else while rebuilding the app.

I was thinking to build an editor (Emacs) plugin next which would do instant push on saving a file. I think most editors can keep a process running in the background. It's really easy to develop an editor plugin for hot-pushing to meteor. All you have to do is start the meteor-hot-loader process, and push code to it with POST requests whenever a file changes.

I hope I'll get some time to do these soon.
